### ISSUE TITLE
sub folder routing doesn't seem to be working.

### ISSUE THREAD

**Issue by PeterClough** [ngbp/ngbp#295]
Hi all 

I've been a php developer for quite a while and have decided to make the leap to JS. So I've been hunting around I found AngularJS and immediately saw the awesomeness etc and then saw Strongloop was a good back end solution. Great so I cloned this https://github.com/carlodicelico/loopback-ng-boilerplate which of course includes ng-boilerplate.

Just been having a tentative play with it and have been trying to get the routing working through subfolders.

```
/app
    /users
    /users.js
        /login
        /login.js
```

users.js

```
angular.module( 'gardeners.users', [
    'ui.router',
    'placeholders',
    'ui.bootstrap',
    'users.login'
])

.config(function config( $stateProvider ) {
    $stateProvider.state( 'users', {
        url: '/users',
        views: {
            "main": {
                controller: 'UsersCtrl',
                templateUrl: 'users/users.tpl.html'
            }
        },
        data:{ pageTitle: 'What is It?' }
    });
})

.controller( 'UsersCtrl', function UsersCtrl( $scope ) {
    // This is simple a demo for UI Boostrap.
    $scope.dropdownDemoItems = [
        "The first choice!",
        "And another choice for you.",
        "but wait! A third!"
    ];
})

;
```

login.js

```
angular.module( 'users.login', [
    'ui.router',
    'placeholders',
    'ui.bootstrap'
])

.config(function config( $stateProvider ) {
    $stateProvider.state( 'usersLogin', {
        url: '/users/login',
        views: {
            "main": {
                controller: 'UsersLoginCtrl',
                templateUrl: 'users/login/login.tpl.html'
            }
        },
        data:{ pageTitle: 'What is It?' }
    });
})

.controller( 'UsersLoginCtrl', function UsersLoginCtrl( $scope ) {
    // This is simple a demo for UI Boostrap.
    $scope.dropdownDemoItems = [
        "The first choice!",
        "And another choice for you.",
        "but wait! A third!"
    ];
})

;
```

my app.js is

```
angular.module( 'gardeners', [
  'templates-app',
  'templates-common',
  'gardeners.home',
  'gardeners.about',
  'gardeners.users',
  'ui.router',
  'lbServices'
])

.config( function myAppConfig ( $stateProvider, $urlRouterProvider ) {
  $urlRouterProvider.otherwise( '/home' );
})

.run( function run () {
})

.controller( 'AppCtrl', function AppCtrl ( $scope, $location ) {
  $scope.$on('$stateChangeSuccess', function(event, toState, toParams, fromState, fromParams){
    if ( angular.isDefined( toState.data.pageTitle ) ) {
      $scope.pageTitle = toState.data.pageTitle + ' | gardeners' ;
    }
  });
})

;

```

The straight 'users' url works but the /users/login fails with this error in the console 

```
Could not resolve 'users/login' from state ''
```

If I debug it I can see it error out on the $stateProvider line in login.js when it loads.

Anyone tell me what schoolboy error or what I'm misunderstanding I've made, would be very grateful thanks.

**Comment by PaulL1** [comment]
I can’t see it by just looking, but I reckon there’s a reasonable chance it’s because of the nesting somehow.  I’d try to get it working as /login, then move it to /users/login.

Paul Lambert
paul@planar.id.au
+61 411 385 488

**Comment by PaulL1** [comment]
I’ve always seen child states defined with dot notation - so “users.login.”  The doco discusses it: https://github.com/angular-ui/ui-router/wiki/Nested-States-%26-Nested-Views, and I think it says that if you don’t use dot notation then you have to explicitly declare the parent.  I can’t recall whether you can have the folders nested without similarly nesting the state - it seems like you should be able to, but perhaps you can’t.

Paul Lambert
paul@planar.id.au
+61 411 385 488

**Comment by Merott** [comment]
As @PaulL1 mentioned, nested states are generally defined using the [dot notation](https://github.com/angular-ui/ui-router/wiki/Nested-States-%26-Nested-Views#dot-notation), so in your case, `usersLogin` should be `users.login`. Alternatively you can explicitly define the parent relationship using the [parent property](https://github.com/angular-ui/ui-router/wiki/Nested-States-%26-Nested-Views#parent-property-using-state-name-string).

Also note that with hierarchical states, the `url` property of the state should be relative to its parent, so for your child `users.login` state, the `url` would be simply `/login`.

Lastly, the directory structure has no effect on the routing. You can set up your directory structure however you like - as long as the `ngbp` build system can find them.

**Comment by PeterClough** [comment]
Thanks guys I have changed the following two lines but still the same problem

```
    $stateProvider.state( 'users.login', {
        url: '/login',
```

I'll carry on digging this is weird.

**Comment by joshdmiller** [comment]
I agree with @Merott and @PaulL1 about nested routes, however I'm not sure @PeterClough wants that here.

It's very hard to tell without being able to play with the code, but for the record I've used very deeply nested routes without issue. Can you push your code up to github so I can pull it down and take a look?

But before you do, try changing your two route URLs to these:

``` javascript
  $stateProvider.state( 'users', {
        url: '^/users',
  // ...
  $stateProvider.state( 'usersLogin', {
        url: '^/users/login',
  // ...
```

The carrot will ensure it matches [absolute routes](https://github.com/angular-ui/ui-router/wiki/URL-Routing#absolute-routes-).

**Comment by PeterClough** [comment]
I found what I was doing wrong, I've literally just started playing with all this. I realised much later 2 things that weren't correct. Firstly and very embarrassingly I had put the url in the index.html not the state :/. Secondly I had no ng-view tag in the parent html snippet, odd things seemed to happen until I made the ng-view specific as well ie. ng-view="users". To add to all that confusion I was messing about deleting and adding child states with grunt watch running and my template-app.js was completely screwed, so I deleted that and ran a grunt compile and everything worked fine.

Thanks for all the help.
